Dip/src/AlpsDecompTreeNode.cpp

```diff
@@ -305,8 +305,7 @@ int AlpsDecompTreeNode::process(bool isRoot,
       //watch tolerance here... if quality is close enough, fathom it
       gap = UtilCalculateGap(thisQuality, currentUB, decompAlgo->getInfinity());
 
-      //if(gap <= relTolerance){
-      if (quality_ >= currentUB) {
+      if(gap <= relTolerance){
```

Comment:
I guess that `MasterGapLimit` is meant for a slightly different purpose---to determine when the upper and lower bounds are "close enough" that the column generation can be considered to have converged. What is being done here is to say that the node can be fathomed as long as the best lower bound is within a certain tolerance of the global upper bound. The current check (`quality_ >= currentUB`) requires the lower bound to exceed the upper bound with no tolerance and is thus just a bit more conservative. Realistically, I guess this change won't make much difference, but the more conservative check seems to make sense to me. What is your thinking? 

A middle ground would be to use a different parameter that is just for this purpose and set the default to 0. Then you could change it if you want.

Reply:
The parameter is probably the wrong one. I think perhaps it should be `TolZero` instead. Not sure maybe something else. And there is also a `primalTolerance` being used here https://github.com/coin-or/Dip/blob/0f856ec34e79aa7c30b7b6c3599ef5ba530174ed/Dip/src/AlpsDecompTreeNode.cpp#L147

Both tolerance params are set as constants
https://github.com/coin-or/Dip/blob/0f856ec34e79aa7c30b7b6c3599ef5ba530174ed/Dip/src/AlpsDecompTreeNode.cpp#L127
https://github.com/coin-or/Dip/blob/0f856ec34e79aa7c30b7b6c3599ef5ba530174ed/Dip/src/AlpsDecompTreeNode.cpp#L134

Regarding if the check should be changed at all. I was solving a linear MCF and got numerical instability with the UB so it was something like 12.00000000001 and the current node quality (LB) was 12. Then branching was started although I was solving an LP. I noticed that when checking bounds nodes further down the branch tree it is done using the a tolerance https://github.com/coin-or/Dip/blob/0f856ec34e79aa7c30b7b6c3599ef5ba530174ed/Dip/src/AlpsDecompTreeNode.cpp#L258 
So my idea was to have similar behavior for root and not root nodes in the branch tree and therefore allow some tolerance in the gap.

I think there should be one param for something being close to zero. For Alps it could even be the Alps settings https://github.com/coin-or/CHiPPS-ALPS/blob/656b1976d7b3e56be1883d72b87981cc406a5df4/Alps/src/Alps.h#L299-L301

Reply:
Hmm, I see what you're saying, but I'm not sure a tolerance is a way to solve this particular issue. Branching should not be invoked whenever the solution obeys integrality constraints, regardless of bounds. If you are solving an LP, then there are not integrality constraints so branching shouldn't happen. Can you point out the code path where this can happen or send me an instance to replicate? 

SYMPHONY actually goes in the opposite direction and won't prune unless the lower bound exceeds the upper by some small positive tolerance and I have never observed this to cause any problems. Pruning when the lower bound hasn't quite reached the upper bound seems problematic.

In any case, making the amount by which the lower bound must exceed the upper bound a parameter that you could set to some small negative value if you really want to is fine, but I would set the default to zero or even some small positive value.  

Could we take the opportunity to try to clean up all the tolerances and make a small set of them that are all parameters? I guess they could be something like:
- zero tolerance (comparison to zero)
- integer tolerance (how close to integer the value of an integer variable should be)
- optimality tolerance (how close the global lower bound must be to the global upper bound before solution is declared optimal---relative or absolute)
- LP error tolerance (tolerance for error in the LP objective function value)
By these parameters, one could always prune as long as the relative gap for an individual subproblem is less than or equal to the relative optimality tolerance, but one also has to take into account the possible error in the LP solution value. So one could maybe require 
```
relGap <= relOptTol
```
where `relGap` is computed according `lower bound - lpTol` or something like that.

By the way, we could also compute a granularity based on the gcd of the objective function coefficients, which would allow earlier pruning. In other words, if the objective function is integer, then we can prune as soon as the lower bound is within the upper bound - 1.